wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -379,54 +381,50 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
     startingPoint <- case start of
         Nothing -> withNode $ getFirstGenesisBlockHash genesisHash
         Just sh -> nextHistoricalHash sh >>= maybe (throwM $ RestorationSuccessorNotFound sh) pure
-    restore genesisHash startingPoint NoTimingData
+    restore genesisHash startingPoint 1000
   where
     wId :: WalletId
     wId = WalletIdHdRnd rootId
 
     -- Process the restoration of the block with the given 'HeaderHash'.
-    restore :: GenesisHash -> HeaderHash -> TimingData -> IO ()
-    restore genesisHash hh timing = do
-
-        -- Updating the average rate every 5 blocks.
-        (rate, timing') <- tickTiming 5 timing
-
-        -- Update each account's historical checkpoints
-        block <- getBlockOrThrow genesisHash hh
+    restore :: GenesisHash -> HeaderHash -> Int -> IO ()
+    restore genesisHash hh batchSize = do
+        -- Keep track of timing
+        startTime <- getCurrentTime
+
+        -- Processing a batch of blocks, flushing to Acid-State only once
+        (updates, (hh', finished)) <- flip runStrictStateT (hh, False) $ sequence $ replicate batchSize $ do
+            (currentHash, finished) <- get
+            if finished then return Nothing else do
+                if tgtHash == currentHash then
+                    put (currentHash, True)
+                else
+                    lift (nextHistoricalHash currentHash) >>= \case
+                        Nothing  -> throwM (RestorationFinishUnreachable tgtHash currentHash)
+                        Just hh' -> put (hh', False)
+                lift $ getPrefilteredBlockOrThrow genesisHash currentHash
+
+        -- Flush to Acid-State & SQLite
+        k <- getSecurityParameter (wallet ^. walletNode)
+        let (blocks, txMetas, slotIds) = unzip3 (catMaybes updates)
+        mErr <- update (wallet ^. wallets) (ApplyHistoricalBlocks k rootId blocks)
+        whenLeft mErr (throwM . RestorationApplyHistoricalBlockFailed)
+        forM_ (mconcat txMetas) (putTxMeta (wallet ^. walletMeta))
```

Comment:
Why is this not another instance of the same problem, with multiple sqlite DB transactions? Is it because this only covers txs that actually belong to our wallet, rather than having an update per-block irrespective?

Reply:
I don't think this is a major concern, since most blocks contain no transactions that are relevant to the wallet, and so this won't do anything. But it might indeed make sense to introduce a `putTxMetas` call; it would be trivial to use here.